src/Kaleidoscope.h

```diff
@@ -57,7 +165,228 @@ class KaleidoscopePlugin {
+  };
+
+ public:
+
+  // The following callbacks handle the synchronized communication
+  // between the Kaleidoscope core and its plugins.
+  //
+  // A plugin can, but does not have to implement all provided hook methods.
+  // Calls to unimplemented hook methods are entirely optimized away by the compiler.
+  // Thus, a non-implemented hook causes neither memory consumption nor
+  // run-time overhead.
+  //
+  // Note: All hook methods in this class are non virtual on purpose as the actual
+  //       interface between the Kaleidoscope singleton class and the
+  //       plugins is the polymorphic PluginHookAdapter class
+
+  void init() {
+    // By letting the new init method call the legacy
+    // begin() method, we make sure that the old hooking
+    // interface is supported even if KALEIDOSCOPE_INIT_PLUGINS()
+    // and KALEIDOSCOPE_CONNECT_PLUGINS are used to register
+    // a plugin that relies on the legacy begin() method
+    // to initialize itself and register hooks.
+    //
+    this->begin();
+  }
+
+  // This handler is supposed to return false if no other handlers are
+  // supposed to be called after it, false otherwise.
+  //
+  // The handler is allowed to modify the mappedKey that is therefore
+  // passed by reference.
+  //
+  bool eventHandlerHook(Key &mappedKey, const EventKey &eventKey) {
+    return true; // Always allow other handlers to continue
+  }
+
+  void preReportHook() {}
+  void postReportHook() {}
+};
+
+// The following invoke macros are meant to be used with
+// KALEIDOSCOPE_CONNECT_PLUGINS in conjunction with the MAP macro
+// that casts a specific operation on every member of
+// a variadic macro argument list (if present).
+//
+#define INVOKE_HOOK_FOR_PLUGIN(PLUGIN) \
+__NN__   \
+__NN__   hook_return_val = Hook__::invoke(PLUGIN, hook_args...); \
+__NL__   \
+__NL__   if (!ContinuationPredicate__::eval(hook_return_val)) { \
+__NL__      return hook_return_val; \
+__NL__   }
+
+#define INVOKE_EMPTY_ARGS_HOOK_FOR_PLUGIN(PLUGIN) \
+   \
+   Hook__::invoke(PLUGIN);
+
+// KALEIDOSCOPE_INIT_PLUGINS is meant to be invoked at global scope of
+// the firmware sketch.
+//
+// Arguments: A list of references to plugin instances that have been
+//       instanciated at global scope.
+//
+// A note concerning possible future optimizations:
+//    In C++17 the loop over all plugins, and calling a hook method
+//    with a given list of arguments can be implemented much more elegant
+//    without using preprocessor macros.
+//    The alternative solution would consist in a combination
+//    of std::forward_as_tuple and std::apply to call a function
+//    (the hook method) for every member of a tuple (containing rvalue
+//    references to all plugins). Thus, both the list of plugins and
+//    and the hook method call arguments (both variadic) could be wrapped
+//    in rvalue-reference-tuples and be passed to some kind
+//    of tuple-for-each algorithm.
+//
+#define KALEIDOSCOPE_INIT_PLUGINS(...) \
```

Comment:
I would find this (and the rest of the TMP code) much more readable if the backslashes were column-aligned to the right of the widest line. It would also be nice if the indentation level was the same as for the rest of the code.